### PR TITLE
Adding Xena Data Option

### PR DESCRIPTION
**Pull request by gwaygenomics**
With this option, controlled access is no longer required. The pull request incorporates the download, processing, and usage of publicly available data from the UCSC Xena Database.

### CODE REVIEW

File: subset_datasets.py [hunk 1]
```diff
@@ -83,10 +85,19 @@
...
 
+# Process Xena mutations
+xena_mut_df = (xena_mut_df.query("effect != 'Silent'")
+               .groupby(['#sample', 'chr', 'gene'])
+               .apply(len).reset_index()).rename(columns={0: 'mutation'})
+
+xena_mut_df = (xena_mut_df.pivot_table(index='#sample', columns='gene',
+               values='mutation', fill_value=0) .astype(bool).astype(int))
+
 # Write subsets to file
```
Comment (dhimmel) [inline subset_datasets.py]: `header=True` is default for `pandas.DataFrame.to_csv`. Did you mean something else?

File: download_xena_data.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+if not os.path.exists('data/xena/mutation-matrix.bz2'):
+    request.urlretrieve(mutation, os.path.join('data', 'xena',
+                                               'mutation-matrix.bz2'))
+
+samples = 'https://ndownloader.figshare.com/files/5864865'
+
+if not os.path.exists('data/xena/samples.tsv'):
+    request.urlretrieve(samples, os.path.join('data', 'xena', 'samples.tsv'))
+
+gene_map = 'https://raw.githubusercontent.com/cognoma/cancer-data/master/'\
```
Comment (dhimmel) [inline download_xena_data.py]: A bit dangerous to use versioned data links and then an unversioned URL here. Protip: press `y` on a GitHub page to switch to the versioned URL.

Going to reference https://github.com/cognoma/cancer-data/issues/23 here -- a gene information table that got uploaded to figshare would make your life easier.

File: pancancer_classifier.py [hunk 3]
```diff
@@ -134,17 +154,37 @@ def get_threshold_metrics(y_true, y_pred, tissue='all'):
...
+                                     .disease.replace(to_replace=ac))
+    base_add = 'XENA'
+else:
+    rnaseq_df.drop('SLC35E2', axis=0, inplace=True)
+    base_add = 'synapse'
 
 # Generate file names for output
-base_fh = 'tissues_' + args.tissues.replace(',', '_') + '_genes_' + \
-          args.genes.replace(',', '_')
+base_fh = base_add + '_tissues_' + args.tissues.replace(',', '_') + \
```
Comment (dhimmel) [inline pancancer_classifier.py]: I like this approach:

```python
'{}_tissues_{}_genes_{}'.format(
    base_add,
    args.tissues.replace(',', '_'),
    args.genes.replace(',', '_')
}
```

While not much simpler ATM, it is good for your training with Python 3.6 around the corner and the ability to use the advanced [Format Specification Mini-Language](https://docs.python.org/3.4/library/string.html#format-specification-mini-language).

Comment (dhimmel) [inline pancancer_classifier.py]: Python 3.6 will change everything.

Coming soon to a terminal near you.

PEP 498 -- Literal String Interpolation

2016-12-16

File: download_xena_data.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# does not have restricted use
+
+import os
+from urllib import request
+from subprocess import call
+
+if not os.path.exists('data/xena'):
+    os.makedirs('data/xena')
+
+expression = 'https://ndownloader.figshare.com/files/5864859'
```
Comment (gwaygenomics) [inline download_xena_data.py]: Nice! I will update

File: download_xena_data.py (hunk 2 above)
Comment (gwaygenomics) [inline download_xena_data.py]: Yes, it would make it easier. Thanks for the versioned URL tip!

File: pancancer_classifier.py (hunk 3 above)
Comment (gwaygenomics) [inline pancancer_classifier.py]: wow! That's great, ok thanks. I will adopt this format for ease in transitioning to mini language

File: subset_datasets.py (hunk 1 above)
Comment (gwaygenomics) [inline subset_datasets.py]: Nope, I guess that snuck in

**Pull request merged**